### PR TITLE
Paper: Reproducible Machine Learning Workflows for Scientists

### PR DESCRIPTION
**Pull request by matthewfeickert**
This PR adds the `myst.md` paper source for the SciPy 2025 proceedings for the tutorial "[Reproducible Machine Learning Workflows for Scientists with Pixi](https://cfp.scipy.org/scipy2025/talk/GDN8PN/)".

This is currently in draft mode as this is being opened in advance of the 

https://github.com/scipy-conference/scipy_proceedings/blob/e53c3dfe5b9813b597fd08f949049d7b250c1808/README.md?plain=1#L108

deadline to ensure that a draft gets in. This contribution will remain silent for most of the time leading up to the deadline as it will be developed in a personal repository where all authors can more easily work on it together.

### CODE REVIEW

File: papers/matthew_feickert/myst.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
+version: 1
+extends: ../papers.yml
+project:
+  # Update this to match `scipy-2025-<folder>` the folder should be `<firstname_surname>`
+  id: scipy-2025-matthew_feickert
```
Comment (matthewfeickert) [inline papers/matthew_feickert/myst.yml:5]: @scipy-conference/2025-proceedings Do you know why my `id` is generating the following error

```
 {
  enforceSingleFolder: true,
  paths: [ 'papers/matthew_feickert' ],
  pathIds: { 'papers/matthew_feickert': 'scipy-2025-matthew_feickert' },
  idPatternRegex: '^scipy-2025-(?:[a-zA-Z0-9-_]{3,15})$',
  changedFiles: [
    'papers/matthew_feickert/acknowledgements.md',
    'papers/matthew_feickert/banner.png',
    'papers/matthew_feickert/conda-forge.md',
    'papers/matthew_feickert/cuda.md',
    'papers/matthew_feickert/introduction.md',
    'papers/matthew_feickert/main.md',
    'papers/matthew_feickert/mybib.bib',
    'papers/matthew_feickert/myst.yml',
    'papers/matthew_feickert/pixi.md'
  ],
  filteredPaths: [ 'papers/matthew_feickert' ],
  unknownChangedFiles: [],
  prLabels: [ 'paper', 'draft' ],
  previewLabel: [ 'draft' ],
  doPreview: true,
  submitLabel: [ 'ready' ],
  doSubmit: false,
  assignees: [],
  reviewers: []
} 


Error: The project IDs are not valid or are not unique, check the error logs for more information.

Invalid id for path "papers/matthew_feickert" (ID: `scipy-2025-matthew_feickert`):
 - Must not be null or empty
 - Must match id-pattern-regex: /^scipy-2025-(?:[a-zA-Z0-9-_]{3,15})$/
 - Only includes "a-z A-Z 0-9 - _"
Update papers/matthew_feickert/myst.yml to include a valid project.id
```

I'm following the 

> Update this to match `scipy-2025-<folder>` the folder should be `<firstname_surname>`

directions.

Comment (rowanc1) [reply papers/matthew_feickert/myst.yml:5]: Extended the ID length in #1087. Can you rebase this PR, and then it should work.

Comment (matthewfeickert) [reply papers/matthew_feickert/myst.yml:5]: Thanks very much @rowanc1!

**Comment by ameyxd** [comment]
Inviting reviewers: @Gift-Ojeabulu and @Krishsan1963@gmail.com

**Comment by matthewfeickert** [comment]
Comment to reviewers: I'll move the edits I've been preparing elsewhere here tonight, so don't be alarmed that there's nothing to review right at this moment!

@ameyxd It seems that the second reviewer handle was their email, not their GitHub user name.

**Comment by Gift-Ojeabulu** [comment]
Seen I will get on that soon.

sorry for the delay,

**Comment by sanjaybk7** [comment]
Commenting to confirm. Will review this week and provide feedback at the earliest. Thanks.

**Comment by ameyxd** [comment]
Hi folk, I will serve as the editor for this paper when reviews are complete.

Editor: @ameyxd

**Comment by Gift-Ojeabulu** [comment]
I am currently reviewing this and will give detailed feedback tomorrow

**Comment by matthewfeickert** [comment]
@Gift-Ojeabulu Sorry, I hadn't added the revised paper. I'll do that tonight, but I unfortunately am fully engaged for the full day.

**Comment by Gift-Ojeabulu** [comment]
I understand that the conference is currently going on, take your time .

I will keep checking.

I hope to get it soon.

**Comment by matthewfeickert** [comment]
@Gift-Ojeabulu @sanjaybk7 I've added some revisions and rebased to get any changes from the upstream `2025` branch. I realized that I left this in "draft" mode for the last weeks, so you might have not yet looked at it. If you have any questions please let me know.

**Comment by ameyxd** [comment]
@sanjaybk7 - can you confirm your review here ASAP?

**Comment by ameyxd** [comment]
@Gift-Ojeabulu - can you confirm your review here, or we would need to reassign reviewers.

File: papers/matthew_feickert/introduction.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+We define this as be limited to the ability to define and programmatically create a software environment composed of packages that specifies all software, and its dependencies, with exact URLs and binary digests ("hashes").
+Reproducible environments need to be machine agnostic in that for a specified computing platform in the environment they must be installable without modification across multiple instances.
+
+### Hardware accelerated environments
+
+Software the involves hardware acceleration on computing resources like GPUs requires additional information to be provided for full computational reproducibility.
+In addition to the computer platform, information about the hardware acceleration device, its supported drivers, and compatible hardware accelerated versions of the software in the environment (GPU enabled builds) are required.
+Traditionally this has been very difficult to do, but multiple recent technological advancements (made possible by social agreements and collaborations) in the scientific open source world now provide solutions to these problems.
```
Comment (ameyxd) [inline papers/matthew_feickert/introduction.md]: Please mention why this has been difficult, concisely.

File: papers/matthew_feickert/pixi.md [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+Importantly, Pixi automatically and non-optionally will produce or update a lock file &mdash; a structured file that contains a full list of all environments defined with a complete list of all packages, as well as definition of each packages with digest information on the binary &mdash; for the software environments defined by the user whenever any actions mutate the environment.
+Pixi is written in Rust, and leverages the language's speed and technologies to solve environments fast.
+
+Pixi addresses the concept of computational reproducibility by focusing on a set of main features
+
+1. **Virtual environment management**: Pixi can create environments that contain conda packages and Python packages and use or switch between environments easily.
+1. **Package management**: Pixi enables the user to install, update, and remove packages from these environments through the `pixi` command line.
+1. **Task management**: Pixi has a task runner system built-in, which allows for tasks with custom logic and dependencies on other tasks to be created.
+
+combined with robust behaviors
```
Comment (ameyxd) [inline papers/matthew_feickert/pixi.md]: Is this a stray phrase?

File: papers/matthew_feickert/pixi.md [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+This ensures that any and every state of a Pixi project is trivially computationally reproducible.
+1. **Solving environments for other platforms**: Pixi allows the user to solve environment for platforms other than the current user machine's.
+This allows for users to solve and share environment to any collaborator with confidence that all environments will work with no additional setup.
+1. **Pairity of conda and Python packages**: Pixi allows for conda packages and Python packages to be used together seamlessly, and is unique in its ability to handle overlap in dependencies between them.
+Pixi will first solve all conda package requirements for the target environment, lock the environment, and then solve all the dependencies of the Python packages for the environment, determine if there are any overlaps with the existing conda environment, and the only install the missing Python dependencies.
+This ensures allows for fully reproducible solves and for the two package ecosystems to compliment each other rather than potentially cause conflicts.
+1. **Efficient caching**: Pixi uses an extremely efficient global caching scheme.
```
Comment (ameyxd) [inline papers/matthew_feickert/pixi.md]: Mention why/how it the caching system is efficient here, if possible, or exclude the adverb 'extremely'.

**Review by ameyxd — COMMENTED** [review]
Had to chime in as a reviewer/editor. All-in-all, good write-up, I recommend a few minor edits. Great job.

**Comment by matthewfeickert** [comment]
Thanks @ameyxd! I am in transit currently but I will aim to revise things based off your review by early next week.

**Comment by ameyxd** [comment]
@matthewfeickert Hi, just a reminder that the Final Author Revision Deadline is 9/4/2025. This means you shouldn't be making any changes after this date. If you want to make changes, please do so before this date!

File: papers/matthew_feickert/introduction.md (hunk 2 above)
Comment (matthewfeickert) [reply papers/matthew_feickert/introduction.md]: This is now

```markdown
While this information is straightforward to collect, traditionally this has been difficult to make use of in practice given software access restrictions and the lack of declarative human interfaces for defining relationships between system-level drivers and user software.
Multiple recent technological advancements (made possible by social agreements and collaborations) in the scientific open source world now provide solutions to these problems.
```

File: papers/matthew_feickert/pixi.md (hunk 3 above)
Comment (matthewfeickert) [reply papers/matthew_feickert/pixi.md]: No, it was supposed to be a continuation of the text directly above it (with the enumerated points being treated as punctuated ideas) but I've revised this to now be

```
These features become powerful when combined with robust behaviors
```

File: papers/matthew_feickert/pixi.md (hunk 4 above)
Comment (matthewfeickert) [reply papers/matthew_feickert/pixi.md]: Ruben has added context on what exactly is happening now.

File: papers/matthew_feickert/linux-containers.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
+## Deploying environments to remote compute
+
+Often researchers are running scientific and machine learning workflows on remote computational resources that use batch computing systems (e.g. HTCondor, SLURM).
+For systems with shared filesystems (e.g. SLURM) it is possible to use Pixi workspaces in workflows in a similar manner to local machine (e.g. laptop or workstation).
+Other systems (e.g. HTCondor) do not have a shared filesystem (e.g. HTCondor), requiring that each worker node receive its own copy of the software environment.
```
Comment (matthewfeickert) [inline papers/matthew_feickert/linux-containers.md:5]: This was made more explicit to give more context on why containerization would be needed.

File: papers/matthew_feickert/pixi.md [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+[environments]
+...
+gpu = ["gpu"]
+inference = ["gpu", "inference"]
+lab = ["gpu", "inference", "lab"]
+```
+
+Composing multiple environments from Pixi features allows for separating conceptual steps of scientific analysis into bespoke software environments that contain only the necessary dependencies.
+This allows for each step's environment to be better defined, potentially with radically different or conflicting dependencies from other steps, and for clean separation between interactive and non-interactive ("batch") computing models.
```
Comment (matthewfeickert) [inline papers/matthew_feickert/pixi.md:181]: The `lab` feature and environment were added to show another conceptual example, and the additional text was added to make the use motivations more clear.

**Review by matthewfeickert — COMMENTED** [review]
Additional text beyond typo fixes and revisions asker for by reviewers were added to provide more conceptual clarity. These changes are outlined at a high level in these comments.

File: papers/matthew_feickert/main.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+:::
+
+:::{include} linux-containers.md
+:::
+
+:::{include} summary.md
+:::
+
+:::{include} acknowledgements.md
+:::
```
Comment (matthewfeickert) [inline papers/matthew_feickert/main.md:29]: In the Curvenote preview these show up as "Supporting documents"

Image

Is there a way to have things be included for a better writing focus experience but have them be understood as a single document?

File: papers/matthew_feickert/pixi.md [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+### CUDA hardware accelerated environment creation
+
+Combining the features of modern CUDA `12` conda packages with Pixi's environment management, it is now possible to efficiently manage multiple software environments that can include both hardware accelerated and CPU environments.
+An example Pixi workspace is presented in @pixi-ml-example-workspace
+
+```{literalinclude} code/ml-example/pixi.toml
+:linenos:
+:label: pixi-ml-example-workspace
+:caption: Example of a multi-platform and multi-environment Pixi manifest with all required information and constraints to resolve and install CUDA accelerated conda packages.
+```
```
Comment (matthewfeickert) [inline papers/matthew_feickert/pixi.md:41]: There is an extra line that is included in the render

Image

though the included file only has 50 lines 

```console
$ wc -l papers/matthew_feickert/code/ml-example/pixi.toml 
50 papers/matthew_feickert/code/ml-example/pixi.toml
```

This doesn't happen with the Dockerfile later though, and I'm not sure why.

File: papers/matthew_feickert/pixi.md [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+
+[feature.cpu.tasks.train-cpu]
+description = "Train a PyTorch CNN on MNIST on CPU"
+cmd = "python src/torch_MNIST.py --epochs 2 --save-model --data-dir data"
+
+...
+
+[environments]
+cpu = ["cpu"]
+```
```
Comment (matthewfeickert) [inline papers/matthew_feickert/pixi.md:71]: If I want to show components of a file that I would normally include, but want to show only segments of it (like here where I separate segments with `...`) is it possible to do this with MyST?

**Review by matthewfeickert — COMMENTED** [review]
@scipy-conference/2025-proceedings 👋 I have some formatting questions (probably mostly for the Curvenote team members). I don't think any of them are critical, but for visual quality of the proceedings these would be interesting to know if they can be addressed. (Apologies for not asking these questions 2 months ago. 😬)

File: papers/matthew_feickert/main.md (hunk 7 above)
Comment (rowanc1) [reply papers/matthew_feickert/main.md:29]: If your main document is the one with all the includes, you can edit the `myst.yml` with the `toc:` to _only_ include the `main.md`. Some [docs here](https://mystmd.org/guide/table-of-contents).

Use `myst init --write-toc` and then remove the ones that shouldn't be there!

Hopefully that makes sense!

File: papers/matthew_feickert/pixi.md (hunk 8 above)
Comment (rowanc1) [reply papers/matthew_feickert/pixi.md:41]: I am at a loss. Listed this here #1082, we will fix before publishing.

File: papers/matthew_feickert/pixi.md (hunk 9 above)
Comment (rowanc1) [reply papers/matthew_feickert/pixi.md:71]: I _think_ it might be possible using the `lines` option of `literalinclude` here:
https://mystmd.org/guide/directives#directive-include

However, there isn't the `...` option for sure (I am also not quite sure how the line numbers show up, likely continuous, which might not be what you want.)

Comment (matthewfeickert) [reply papers/matthew_feickert/pixi.md:71]: I am going to be bad and make one change after the "pens down" deadline (https://github.com/scipy-conference/scipy_proceedings/pull/1085#issuecomment-3228823298) as I noticed a path inconsistency between the actual TOML file and the copy pasted examples.

File: papers/matthew_feickert/main.md (hunk 7 above)
Comment (matthewfeickert) [reply papers/matthew_feickert/main.md:29]: Thanks @rowanc1. This was exactly what I needed.

File: papers/matthew_feickert/pixi.md (hunk 9 above)
Comment (matthewfeickert) [reply papers/matthew_feickert/pixi.md:71]: > I _think_ it might be possible using the `lines` option of `literalinclude` here: https://mystmd.org/guide/directives#directive-include
> 
> ... (I am also not quite sure how the line numbers show up, likely continuous, which might not be what you want.)

@rowanc1 Yeah, exactly. This is what I tried at first, but then indeed the line numbers are sequential, which for comparison would be more confusing then having no line numbers.

e.g.

~~~markdown
```{literalinclude} code/ml-example/pixi.toml
:linenos:
:start-line: 34
:end-line: 40
:emphasize-lines: 4-6
```
~~~

gives

image

**Comment by matthewfeickert** [comment]
As @ameyxd had moved this to "Editor signed-off" (thanks!) I'm going to squash all commits and rebase this off of the current `HEAD` of `scipy-conference:2025` to make this easy and a clean commit for the future.

**Review by ameyxd — APPROVED** [review]
LGTM, ready to merge. Approved.

File: papers/matthew_feickert/pixi.md (hunk 8 above)
Comment (fwkoch) [reply papers/matthew_feickert/pixi.md:41]: Looked into this a little bit: there _is_ a trailing new line in `pixi.toml` (and `Dockerfile`). `wc -l` does not count this as an extra line (I also get `50`):
image
(In github you don't see the `⊖ No newline` symbol.)

Using `:linenos:` in `literalinclude`, the trailing newline is picked up; without `:linenos:` (as is the case with `Dockerfile`) it's ignored. All this said, I think the behavior in myst is correct... We could `trim` the included content... but I'd worry that could have unintended consequences, maybe...?

We can definitely fix this with `:end-line:` - in the case of `pixi.toml`, both `:end-line: -2` and `:end-line: 49` work (ugh, here the zero- vs one-indexing is dicey - both of these make `50` the actual end-line).

Anyway! I'll make this change for you now.

File: papers/matthew_feickert/pixi.md [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+
+### CUDA hardware accelerated environment creation
+
+Combining the features of modern CUDA `12` conda packages with Pixi's environment management, it is now possible to efficiently manage multiple software environments that can include both hardware accelerated and CPU environments.
+An example Pixi workspace is presented in @pixi-ml-example-workspace
+
+```{literalinclude} code/ml-example/pixi.toml
+:linenos:
+:label: pixi-ml-example-workspace
+:caption: Example of a multi-platform and multi-environment Pixi manifest with all required information and constraints to resolve and install CUDA accelerated conda packages.
```
Comment (fwkoch) [inline papers/matthew_feickert/pixi.md:40]:
```suggestion
```{literalinclude} code/ml-example/pixi.toml
:linenos:
:end-line: 49
:label: pixi-ml-example-workspace
:caption: Example of a multi-platform and multi-environment Pixi manifest with all required information and constraints to resolve and install CUDA accelerated conda packages.
```

**Comment by matthewfeickert** [comment]
Thanks you for fixing things up @fwkoch (and Curvenote team)!

File: papers/matthew_feickert/pixi.md (hunk 8 above)
Comment (matthewfeickert) [reply papers/matthew_feickert/pixi.md:41]: > Looked into this a little bit: there is a trailing new line in `pixi.toml` (and `Dockerfile`).
> ...
> All this said, I think the behavior in myst is correct...

@fwkoch The point here is that _only_ for the `pixi.toml` include was this not working. There are no issues with including `papers/matthew_feickert/code/ml-example/Dockerfile` or `papers/matthew_feickert/example.lock`, which have the same endline. So why is there different behavior between the the `pixi.toml` and everything else?

Comment (fwkoch) [reply papers/matthew_feickert/pixi.md:41]: Hey @matthewfeickert - in case you are still following along. This is not related to the file, it's whether or not you use the `:linenos:` option. If you use this option, the trailing newline will show up numbered; if you do not, it will be trimmed.

This comes all the way down to the frontend package MyST uses for rendering code - there is an open issue for this exact thing here: https://github.com/react-syntax-highlighter/react-syntax-highlighter/issues/443
